connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/inbound/DefaultProcessInstanceContext.java

```diff
@@ -16,45 +16,69 @@
 import java.util.HashMap;
 import java.util.Map;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
 
-public class DefaultDynamicPropertyManager implements DynamicPropertyManager {
-  private static final Logger LOGGER = LoggerFactory.getLogger(DefaultDynamicPropertyManager.class);
+public final class DefaultProcessInstanceContext implements ProcessInstanceContext {
+  private final Long processInstanceKey;
+  private final ValidationProvider validationProvider;
   private final FeelEngineWrapper feelEngineWrapper;
-  private final Map<Long, Map<String, Object>> propertiesById;
+  private final ObjectMapper objectMapper;
 
-  public DefaultDynamicPropertyManager(
+  private final Map<String, Object> properties;
+
+  public DefaultProcessInstanceContext(
+      final Long processInstanceKey,
+      final ValidationProvider validationProvider,
       final FeelEngineWrapper feelEngineWrapper,
-      final Map<Long, Map<String, Object>> propertiesById) {
+      final ObjectMapper objectMapper,
+      final Map<String, Object> properties) {
+    this.processInstanceKey = processInstanceKey;
+    if (validationProvider == null) {
+      this.validationProvider = ValidationUtil.discoverDefaultValidationProviderImplementation();
+    } else {
+      this.validationProvider = validationProvider;
+    }
     this.feelEngineWrapper = feelEngineWrapper;
-    this.propertiesById = propertiesById;
+    this.objectMapper = objectMapper;
+    this.properties = new HashMap<>(properties);
   }
 
-  public void setOrUpdateProperties(Long id, Map<String, Object> properties) {
-    propertiesById.putIfAbsent(id, new HashMap<>());
-    propertiesById.get(id).putAll(properties);
+  @Override
+  public void update(final Map<String, String> variables) {
+    properties.putAll(variables);
   }
 
-  public Map<String, Object> getProperties(Long id) {
-    return propertiesById.get(id);
+  @Override
+  public Long getProcessInstanceKey() {
+    return processInstanceKey;
   }
 
-  public boolean arePropertiesReadyForBinding(Long propertyKey) {
-    Map<String, Object> properties = propertiesById.get(propertyKey);
+  @Override
+  public <T> T bindProperties(final Class<T> cls) {
+    var mappedObject = objectMapper.convertValue(properties, cls);
+    validationProvider.validate(mappedObject);
+    return mappedObject;
+  }
+
+  @Override
+  public boolean isReadyToBind() {
     for (Map.Entry<String, Object> entrySet : properties.entrySet()) {
       String value = entrySet.getValue().toString();
       if (isFeelExpression(value)) {
         try {
           Object evaluatedValue = feelEngineWrapper.evaluate(value, properties);
+          if (evaluatedValue == null) {
+            return false;
+          }
           entrySet.setValue(evaluatedValue);
         } catch (FeelEngineWrapperException e) {
-          LOGGER.debug("Failed to evaluate FEEL expression", e);
           return false;
         }
       }
```

Comment:
Why is that check needed?

Reply:
I dont see a need for this. You bind agains properties + process variables. We dont pre-validate other bind processes so why do you want to do it here?

Reply:
@Oleksiivanov Why would `headerValue` be null "for now" if we always request processes that have all variables available? The variables should all be there when we execute the bind call.